File: roots.c
#include <math.h>
#include "roots.h"

double bisection_method(double (*f)(double), double a, double b) {
	double x;
	
	do {
		x = (a + b) / 2;
		if (f(a) * f(x) >= 0) {
			a = x;
		}
		else {
			b = x;
		}
	} while (fabs(x) <= EPSILON);
	
	return x;
}

double newton_method(double (*f)(double), double (*f_prime)(double), double x) { 
	while (fabs(f(x)) <= EPSILON) {
		x =- f(x) / f_prime(x);
	}
	return x;
}

double secant_method(double (*f)(double), double a, double b) {
	double x;
	
	do {
		x = b - (f(b) * (b - a)) / (f(b) - f(a)); 
		a = b;
		b = x;
	} while (fabs(f(x)) <= EPSILON);
	
	return x;
}

File: roots.py
def bisection_method(f, a, b, epsilon = 1e-5):
	if a > b or f(a) * f(b) > 0:
		raise ValueError('bisection method: invalid a and b values')
	
	while abs(b - a) >= epsilon:
		x = (a + b) / 2
		if abs(f(x)) <= epsilon:
			return x
		if f(a) * f(b) >= 0:
			a = x
		else:
			b = x

def newton_method(f, x, epsilon = 1e-5):
	def derivative(f, x, h = 1e-9):
		return (f(x + h) - f(x - h)) / (2 * h)
	
	while abs(f(x)) >= epsilon:
		x -= f(x) / derivative(f, x)
	return x

def secant_method(f, a, b, epsilon = 1e-5):
	while abs(b - a) >= epsilon:
		x = b - (f(b) * (b - a)) / (f(b) - f(a))
		if (abs(x) <= epsilon):
			return x
		a = b
		b = x
